runtime/cudaq/builder/kernels.h

```diff
@@ -15,104 +15,111 @@ namespace cudaq {
+std::vector<std::size_t> getControlIndices(std::size_t numControls);
 
 /// @brief Apply a uniformly controlled rotation on the target qubit.
 template <typename Kernel, typename RotationFunctor>
 void applyRotation(Kernel &&kernel, RotationFunctor &&rotationFunctor,
-                   QuakeValue qubits, const std::vector<double> &alpha,
-                   const std::vector<std::size_t> &controls,
-                   std::size_t target) {
-  auto thetas = computeAngle(alpha);
-  auto gcRank = controls.size();
-  if (gcRank == 0) {
+                   QuakeValue qubits, const std::span<double> alphas,
+                   std::size_t numControls, std::size_t target) {
+  auto thetas = convertAngles(alphas);
+  if (numControls == 0) {
     rotationFunctor(kernel, thetas[0], qubits[target]);
     return;
   }
 
-  auto ctrlIds = details::getControlIndices(gcRank);
-  for (auto [i, ctrlIdx] : cudaq::enumerate(ctrlIds)) {
+  auto controlIndices = getControlIndices(numControls);
+  assert(thetas.size() == controlIndices.size());
+  for (auto [i, c] : cudaq::enumerate(controlIndices)) {
     rotationFunctor(kernel, thetas[i], qubits[target]);
-    kernel.template x<cudaq::ctrl>(qubits[controls[ctrlIdx]], qubits[target]);
+    kernel.template x<cudaq::ctrl>(qubits[c], qubits[target]);
   }
 }
 
-/// @brief Return angles required to implement a controlled-Z rotation on
-/// the `kth` qubit.
+/// @brief Return angles required to implement a uniformly controlled z-rotation
+/// on the `kth` qubit.
 std::vector<double> getAlphaZ(const std::span<double> data,
                               std::size_t numQubits, std::size_t k);
 
-/// @brief Return angles required to implement a controlled-Y rotation on
-/// the `kth` qubit.
+/// @brief Return angles required to implement a uniformly controlled y-rotation
+/// on the `kth` qubit.
 std::vector<double> getAlphaY(const std::span<double> data,
                               std::size_t numQubits, std::size_t k);
 } // namespace details
 
-/// @brief Decompose the input state vector data to a set of
-/// controlled operations and rotations. This function takes as input
-/// a `kernel_builder` and appends the operations of the decomposition
-/// to its internal representation. This implementation follows the algorithm
-/// defined in `https://arxiv.org/pdf/quant-ph/0407010.pdf`.
+/// @brief Decompose the input state vector data to a set of controlled
+/// operations and rotations. This function takes as input a `kernel_builder`
+/// and appends the operations of the decomposition to its internal
+/// representation. This implementation follows the algorithm defined in
+/// `https://arxiv.org/pdf/quant-ph/0407010.pdf`.
 template <typename Kernel>
 void from_state(Kernel &&kernel, QuakeValue &qubits,
-                const std::span<std::complex<double>> data,
+                const std::span<std::complex<double>> amplitudes,
                 std::size_t inNumQubits = 0) {
```

Comment:
I think it would be helpful to add some sample amplitude arrays and some expected qubit numbers here. I.e., if `amplitudes[] = {0, 1, 0, 0, 0, 0, 0, 0}`, what would the resulting `sample()` bitstring return? Would it depend on `inNumQubits` or `qubits.constantSize()` was larger than `amplitudes.size()` (i.e. if zero padding were required)?

Reply:
I added a test for checking the qubit ordering. This API will change in a future PR: We should take a `cudaq::state` as input, and we don't need `inNumQubits`. I would not document how qubit ordering and state vector amplitudes are ordered in this function. We to define this in the specification and be consistent across the board.